components/feature/search/src/main/java/mozilla/components/feature/search/widget/AppSearchWidgetProvider.kt

```diff
@@ -0,0 +1,315 @@
+                R.layout.mozac_search_widget_large -> {
+
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab,
+                        textSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_voice,
+                        voiceSearchIntent
+                    )
+                    setOnClickPendingIntent(
+                        R.id.mozac_button_search_widget_new_tab_icon,
+                        textSearchIntent
+                    )
+                    setTextViewText(R.id.mozac_button_search_widget_new_tab, text)
+
+                    // Unlike "small" widget, "medium" and "large" sizes do not have separate layouts
+                    // that exclude the microphone icon, which is why we must hide it accordingly here.
+                    if (voiceSearchIntent == null) {
+                        setViewVisibility(R.id.mozac_button_search_widget_voice, View.GONE)
+                    }
+                }
+            }
+        }
+    }
+
+    private fun RemoteViews.setMicrophoneIcon(context: Context) {
+        setImageView(
+            context,
+            R.id.mozac_button_search_widget_voice,
+            config.searchWidgetMicrophoneResource
+        )
+    }
+
+    private fun RemoteViews.setSearchWidgetIcon(context: Context) {
+        setImageView(
+            context,
+            R.id.mozac_button_search_widget_new_tab_icon,
+            config.searchWidgetIconResource
+        )
+        val appName = context.getString(config.appName)
+        setContentDescription(
+            R.id.mozac_button_search_widget_new_tab_icon,
+            context.getString(R.string.search_widget_content_description, appName)
+        )
+    }
+
+    private fun RemoteViews.setImageView(context: Context, viewId: Int, resourceId: Int) {
+        // gradient color available for android:fillColor only on SDK 24+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            setImageViewResource(
+                viewId,
+                resourceId
+            )
+        } else {
+            setImageViewBitmap(
+                viewId,
+                AppCompatResources.getDrawable(
+                    context,
+                    resourceId
+                )?.toBitmap()
+            )
+        }
+    }
+
+    // Cell sizes obtained from the actual dimensions listed in search widget specs
+    companion object {
+        private const val DP_EXTRA_SMALL = 64
+        private const val DP_SMALL = 100
+        private const val DP_MEDIUM = 192
+        private const val DP_LARGE = 256
+        private const val REQUEST_CODE_VOICE = 1
+
+        /**
+         * It updates AppSearchWidgetProvider size and microphone icon visibility
+         */
+        fun updateAllWidgets(context: Context, widgetClassNameApp: AppSearchWidgetProvider) {
```

Comment:
Looking at this being used in Fenix it wasn't immediately clear why this method needs [a new instance of `AppSearchWidgetProvider`](https://github.com/mozilla-mobile/fenix/pull/26331/files#diff-c9271216658683dbfa5f236fc4da16de698b708287062bd1af1767d2489e240cR94)
```kotlin
AppSearchWidgetProvider.updateAllWidgets(requireContext(), SearchWidgetProvider())
```
--- 

The easiest approach to avoid instantiating `SearchWidgetProvider` if we just need the class it to pass just that.
So this method could become
```kotlin
fun updateAllWidgets(context: Context, clazz: Class<AppSearchWidgetProvider>)
```
and then this would be called with
```kotlin
AppSearchWidgetProvider.updateAllWidgets(requireContext(), SearchWidgetProvider::class.java)
```
--- 

We could also make use of an inline type to have it declared here as
```kotlin
inline fun <reified T: KClass<out AppSearchWidgetProvider>> T.updateAllWidgets(context: Context)
```
and then this would be called with
```kotlin
SearchWidgetProvider::class.updateAllWidgets(requireContext())
```

Reply:
Done